// mongoback/dist/source/utils/options/defaultOptions.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var DEFAULT_CONNECTION_OPTIONS = {
    uri: undefined,
    host: 'localhost',
    port: 27017,
    username: undefined,
    password: undefined,
    authenticationDatabase: undefined,
    replicaSetName: undefined,
    srv: false,
    ssl: false,
    sslCAFile: undefined,
    sslPEMKeyFile: undefined,
    sslPEMKeyPassword: undefined,
    sslCRLFile: undefined,
    sslAllowInvalidCertificates: false,
    sslAllowInvalidHostnames: false,
    authenticationMechanism: undefined,
    gssapiServiceName: undefined,
    gssapiHostName: undefined,
    slaveOk: false,
    readPreference: undefined,
    dbpath: undefined,
    directoryperdb: false,
    journal: false,
    ipv6: false
};
var DEFAULT_EXPORTING_OPTIONS = {
    quiet: false,
    verbose: false,
    type: undefined,
    jsonFormat: undefined,
    jsonArray: false,
    pretty: false,
    query: undefined,
    fields: undefined,
    fieldFile: undefined,
    noHeaderLine: false,
    skip: undefined,
    limit: undefined,
    sort: undefined,
    prependDbName: undefined,
    fileName: undefined,
    filePath: undefined,
    absolutePath: false
};
var DEFAULT_EXPORTED_OPTIONS = {
    all: false,
    databases: [],
    collections: [],
    systemCollections: false,
    throwIfLackOfPermissions: false,
    warnIfLackOfPermissions: false,
    throwIfOneFails: false,
    warnIfOneFails: false
};
var DEFAULT_LOG_OPTIONS = {
    silent: false,
    log: 'base'
};
var DEFAULT_OUT_OPTIONS = {
    outDir: './exported',
    outType: 'deep',
    detailedResult: false
};
exports.DEFAULT_OPTIONS = __assign(__assign(__assign(__assign(__assign({}, DEFAULT_CONNECTION_OPTIONS), DEFAULT_EXPORTING_OPTIONS), DEFAULT_EXPORTED_OPTIONS), DEFAULT_LOG_OPTIONS), DEFAULT_OUT_OPTIONS);
