// mongoback/dist/source/utils/logger/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var ora_1 = __importDefault(require("ora"));
var chalk_1 = __importDefault(require("chalk"));
var Logger = /** @class */ (function () {
    function Logger(options) {
        this.spinners = {};
        if (!options.silent && options.log) {
            if (typeof options.log === 'string') {
                options.log = [options.log];
            }
            this.base = options.log.includes('base');
            this.command = options.log.includes('command');
            this.mongoexport = options.log.includes('mongoexport');
            this.expectedCollections = options.log.includes('expectedCollections');
            this.actualCollections = options.log.includes('actualCollections');
        }
    }
    Logger.prototype.exportingDatabase = function (db) {
        if (this.base) {
            console.log(chalk_1.default.white(db));
        }
    };
    Logger.prototype.exportingCollectionStart = function (db, collection) {
        if (this.base) {
            var spinner = ora_1.default({
                text: chalk_1.default.grey(collection),
                indent: 2,
                spinner: 'dots2'
            }).start();
            this.spinners["" + db + collection] = spinner;
        }
    };
    Logger.prototype.exportingCollectionStop = function (db, collection, succeded) {
        if (this.base) {
            var spinner = this.spinners["" + db + collection];
            if (succeded) {
                spinner.succeed();
            }
            else {
                spinner.fail();
            }
        }
    };
    Logger.prototype.printCommand = function (cmd) {
        if (this.command) {
            var tag = chalk_1.default.keyword('darkturquoise')('[COMMAND]');
            var text = chalk_1.default.keyword('lightslategray')(cmd);
            console.log(tag + " " + text);
        }
    };
    Logger.prototype.printMongoexport = function (log, success) {
        if (this.mongoexport) {
            var tag = success ? chalk_1.default.keyword('limegreen')('[SUCCESS]') : chalk_1.default.keyword('orange')('[ERROR]');
            var text = chalk_1.default.white(log);
            console.log(tag + "\n" + text);
        }
    };
    Logger.prototype.printExpectedCollections = function (expected) {
        if (this.expectedCollections) {
            var tag = chalk_1.default.keyword('yellow')('[COLLECTIONS TO EXPORT]');
            var text = JSON.stringify(expected, null, 2);
            console.log(tag + "\n" + text);
        }
    };
    Logger.prototype.printExportedCollections = function (actual) {
        if (this.actualCollections) {
            var tag = chalk_1.default.keyword('yellow')('[COLLECTIONS EXPORTED]');
            var text = JSON.stringify(actual, null, 2);
            console.log(tag + "\n" + text);
        }
    };
    Logger.prototype.warn = function (message, error) {
        var tag = chalk_1.default.keyword('yellow')('[WANING]');
        var text = chalk_1.default.keyword('orange')(message);
        console.warn(tag + " " + text, error);
    };
    return Logger;
}());
exports.Logger = Logger;
